<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Changelog • Giotto</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Changelog" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-109153024-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-109153024-2');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Giotto</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Start
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/getting_started.html#description">Description</a>
    </li>
    <li>
      <a href="../articles/getting_started.html#workflow-diagram">Workflow</a>
    </li>
    <li>
      <a href="../articles/getting_started.html#howtos">HOWTOs</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Documentation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Datasets
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mouse_seqFISH_cortex_200319.html">mouse seqFISH cortex</a>
    </li>
    <li>
      <a href="../articles/mouse_merFISH_hypoth_preoptic_region_200326.html">mouse hypoth. preopt. region</a>
    </li>
    <li>
      <a href="../articles/mouse_starmap_cortex.html">mouse STARmap cortex</a>
    </li>
    <li>
      <a href="../articles/human_cycif_PDAC_200322.html">human CyCIF PDAC</a>
    </li>
    <li>
      <a href="../articles/mouse_visium_brain_200325.html">mouse Visium brain</a>
    </li>
    <li>
      <a href="../articles/mouse_visium_kidney_200326.html">mouse Visium kidney</a>
    </li>
    <li>
      <a href="../articles/mouse_CODEX_spleen.html">mouse CODEX spleen</a>
    </li>
    <li>
      <a href="../articles/mouse_osmFISH_SScortex.html">mouse osmFISH SScortex</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/faqs.html">FAQ</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/RubD/Giotto">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href='https://github.com/RubD/Giotto/blob/master/NEWS.md'><code>NEWS.md</code></a></small>
    </div>

    <div id="giotto-035-development" class="section level1">
<h1 class="page-header" data-toc-text="0.3.5">
<a href="#giotto-035-development" class="anchor"></a>Giotto 0.3.5 (development)</h1>
<ul>
<li>background images See <a href="../articles/howto_images.html">howto’s</a>
</li>
<li>support for sparse matrices<br>
</li>
<li>PCA can be calculated with the packages irlba (default) or factominer (old default)</li>
<li>complemented PCA with separate functions for a scree plot and jackstraw plot</li>
<li>addition of <strong>readExprMatrix</strong> to read an expression matrix</li>
<li>addition of <strong>addGenesPerc</strong> to add information about genesets (e.g. mitochondrial genes)</li>
<li>several bug fixes</li>
</ul>
</div>
    <div id="giotto-032" class="section level1">
<h1 class="page-header" data-toc-text="0.3.2">
<a href="#giotto-032" class="anchor"></a>Giotto 0.3.2</h1>
<ul>
<li>added voronoi plots to use in spatial plotting. See <a href="../articles/howto_voronoi_plots.html">howto’s</a><br>
</li>
<li>generalized visualization parameters between functions</li>
</ul>
</div>
    <div id="giotto-031" class="section level1">
<h1 class="page-header" data-toc-text="0.3.1">
<a href="#giotto-031" class="anchor"></a>Giotto 0.3.1</h1>
<ul>
<li>(optional) automatic installation of python modules through reticulate:
<ul>
<li>you can provide your preferred python path</li>
<li>the giotto environment can be installed automatic</li>
<li>if you do not provide the python path and do not choose to install the giotto environment, then it will take the default python path<br>
</li>
</ul>
</li>
<li>several bug fixes</li>
<li>several mini-datasets are now included within Giotto for quick testing:
<ul>
<li>field 1 of seqFISH+ (single-cell)</li>
<li>the visium brain Dentate Gyrus subset (spots)</li>
<li>subset of starMAP (3D)</li>
</ul>
</li>
</ul>
<p>example to acces the seqFISH+ mini dataset:</p>
<div class="sourceCode" id="cb1"><pre class="r"><span class="co"># raw counts</span>
<span class="no">small_seqfish_expr_matrix</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"seqfish_field_expr.txt"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'Giotto'</span>))
<span class="co"># cell locations</span>
<span class="no">small_seqfish_locations</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"seqfish_field_locs.txt"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'Giotto'</span>))</pre></div>
</div>
    <div id="giotto-030" class="section level1">
<h1 class="page-header" data-toc-text="0.3.0">
<a href="#giotto-030" class="anchor"></a>Giotto 0.3.0</h1>
<ul>
<li>Default spatial network created with <strong>createSpatialNetwork</strong> is now a Delaunay spatial network.</li>
</ul>
<!-- end list --><div class="sourceCode" id="cb2"><pre class="r"><span class="co"># to create the old default kNN spatial network use:</span>
<span class="fu"><a href="../reference/createSpatialKNNnetwork.html">createSpatialKNNnetwork</a></span>(<span class="no">gobject</span>)

<span class="co"># or use this function with the following setting</span>
<span class="fu"><a href="../reference/createSpatialNetwork.html">createSpatialNetwork</a></span>(<span class="no">gobject</span>, <span class="kw">method</span> <span class="kw">=</span> <span class="st">'kNN'</span>)</pre></div>
<ul>
<li>The function names for extracting spatial genes have changed:</li>
</ul>
<!-- end list --><div class="sourceCode" id="cb3"><pre class="r"><span class="co"># binGetSpatialGenes is now:</span>
<span class="fu"><a href="../reference/binSpect.html">binSpect</a></span>(<span class="no">gobject</span>) <span class="co"># binary Spatial extraction</span>

<span class="co"># spatial_genes_python is now:</span>
<span class="fu"><a href="../reference/silhouetteRank.html">silhouetteRank</a></span>(<span class="no">gobject</span>)</pre></div>
<ul>
<li>Fixed multiple bugs</li>
<li>Improved speed by changing code to Rcpp and implementing parallelization options</li>
<li>updated <a href="../articles/getting_started.html">howto’s</a> tutorials in Start section</li>
<li>Finished the analysis of 10 different spatial datasets (tutorials are a work-in-progress)</li>
</ul>
</div>
    <div id="giotto-024" class="section level1">
<h1 class="page-header" data-toc-text="0.2.4">
<a href="#giotto-024" class="anchor"></a>Giotto 0.2.4</h1>
<ul>
<li>New examples on mouse kidney and brain using the recently released <a href="https://www.10xgenomics.com/spatial-transcriptomics/">10X Visium datasets</a> (<strong>NEW</strong>)</li>
<li>Added tools to identify spatial enrichment based on cell-type specific gene signature lists (<strong>NEW</strong>)</li>
</ul>
</div>
    <div id="giotto-023" class="section level1">
<h1 class="page-header" data-toc-text="0.2.3">
<a href="#giotto-023" class="anchor"></a>Giotto 0.2.3</h1>
<ul>
<li>New example with 3D-like spatial data of the mouse hypothalamic preoptic region using <a href="https://science.sciencemag.org/content/362/6416/eaau5324">merFISH</a> (<strong>NEW</strong>)<br>
</li>
<li>New example with 3D spatial data <a href="https://science.sciencemag.org/content/361/6400/eaat5691">STARmap</a>
</li>
<li>New example with the highly sensitive data from <a href="https://www.nature.com/articles/s41592-018-0175-z">osmFISH</a>
</li>
<li>New example on the Cerebellum with the scalable data from <a href="https://science.sciencemag.org/content/363/6434/1463">Slideseq</a>
</li>
<li>New example on mouse olfactory bulb using immobilized primers on glass slides from <a href="https://science.sciencemag.org/content/353/6294/78">Spatial Transcriptomics</a>
</li>
<li>Updated seqFISH+ cortex example (<strong>NEW</strong>)</li>
<li>Updated STARmap cortex example (<strong>NEW</strong>)</li>
</ul>
</div>
    <div id="giotto-022" class="section level1">
<h1 class="page-header" data-toc-text="0.2.2">
<a href="#giotto-022" class="anchor"></a>Giotto 0.2.2</h1>
<ul>
<li>Implemented <a href="https://github.com/Teichlab/SpatialDE">SpatialDE</a> and <a href="https://github.com/edsgard/trendsceek">trendsceek</a>
</li>
<li>Updated support for 3D spatial data<br>
</li>
<li>Added support for the use of global instructions and automatically saving your plots (<strong>NEW</strong>)</li>
<li>Add wrapper for differential expression with <a href="https://github.com/RGLab/MAST">MAST</a> and <a href="https://bioconductor.org/packages/release/bioc/html/scran.html">SCRAN</a>
</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://www.rubendries.com/'>Ruben Dries</a>, Huipeng Li, Qian Zhu, Rui Dong, Guo-Cheng Yuan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


