<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Spatial Single-Cell Transcriptomics Toolbox • Giotto</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Spatial Single-Cell Transcriptomics Toolbox">
<meta property="og:description" content="Toolbox to process, analyze and visualize spatial single-cell expression data.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-109153024-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-109153024-2');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">Giotto</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Start
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/getting_started.html#description">Description</a>
    </li>
    <li>
      <a href="articles/getting_started.html#workflow-diagram">Workflow</a>
    </li>
    <li>
      <a href="articles/getting_started.html#howtos">HOWTOs</a>
    </li>
  </ul>
</li>
<li>
  <a href="reference/index.html">Documentation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Datasets
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/mini_seqfish.html">mini seqFISH</a>
    </li>
    <li>
      <a href="articles/mini_visium.html">mini visium (spots)</a>
    </li>
    <li>
      <a href="articles/mini_starmap.html">mini 3D STARmap</a>
    </li>
    <li>
      <a href="articles/mouse_seqFISH_cortex_200914.html">mouse seqFISH cortex</a>
    </li>
    <li>
      <a href="articles/mouse_merFISH_preoptic_region_200909.html">mouse hypoth. preopt. region</a>
    </li>
    <li>
      <a href="articles/mouse_starmap_cortex_200917.html">mouse STARmap cortex</a>
    </li>
    <li>
      <a href="articles/human_cycif_PDAC_200916.html">human CyCIF PDAC</a>
    </li>
    <li>
      <a href="articles/mouse_visium_brain_201226.html">mouse Visium brain</a>
    </li>
    <li>
      <a href="articles/mouse_visium_kidney_200916.html">mouse Visium kidney</a>
    </li>
    <li>
      <a href="articles/mouse_CODEX_spleen_200921.html">mouse CODEX spleen</a>
    </li>
    <li>
      <a href="articles/mouse_osmFISH_SScortex_200915.html">mouse osmFISH SScortex</a>
    </li>
  </ul>
</li>
<li>
  <a href="articles/contribute.html">Contribute</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    FAQ
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/faqs.html">Frequently Asked Questions</a>
    </li>
    <li>
      <a href="articles/github_issues.html">GitHub Issues Guidelines</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Try Giotto
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/docker.html">Docker</a>
    </li>
    <li>
      <a href="articles/binder.html">Binder</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/RubD/Giotto" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/RubD/Giotto/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="giotto">Giotto<a class="anchor" aria-label="anchor" href="#giotto"></a>
</h1></div>
<!-- badges: start -->
<!-- badges: end -->
<p>The Giotto package consists of two modules, Giotto Analyzer and Viewer (see <a href="http://www.spatialgiotto.com" class="external-link">www.spatialgiotto.com</a>), which provide tools to process, analyze and visualize <strong>single-cell spatial expression</strong> data. The underlying framework is generalizable to virtually all currently available spatial datasets. We recently demonstrated the general applicability on 10 different datasets created by 9 different state-of-the-art spatial technologies, including <em>in situ</em> hybridization (seqFISH+, merFISH, osmFISH), sequencing (Slide-seq, Visium, STARmap) and imaging-based multiplexing/proteomics (CyCIF, MIBI, CODEX). These technologies differ in terms of resolution (single cell vs multiple cells), spatial dimension (2D vs 3D), molecular modality (protein vs RNA), and throughput (number of cells and genes). More information and documentation about the latest <strong>released version</strong> of Giotto Analyzer can be found at <a href="https://rubd.github.io/Giotto_site/" class="external-link uri">https://rubd.github.io/Giotto_site/</a>.</p>
<p><img src="inst/images/general_figs/overview_datasets.png"></p>
<div class="section level2">
<h2 id="requirements">Requirements<a class="anchor" aria-label="anchor" href="#requirements"></a>
</h2>
<ul>
<li>R (&gt;= 3.5.1)</li>
<li>Python (&gt;= 3.0)</li>
<li>Windows, MacOS or Linux specific installation tools. See <a href="https://support.rstudio.com/hc/en-us/articles/200486498-Package-Development-Prerequisites" class="external-link">link</a>.</li>
</ul>
<p> </p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>See <a href="https://rubd.github.io/Giotto_site/articles/faqs.html" class="external-link">FAQs</a> for additional information.</p>
<div class="section level4">
<h4 id="r-installation">R installation<a class="anchor" aria-label="anchor" href="#r-installation"></a>
</h4>
<p>You can install Giotto with (~1-5 mins):</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span>  <span class="co"># if not installed: install.packages('devtools')</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://remotes.r-lib.org" class="external-link">remotes</a></span><span class="op">)</span>  <span class="co"># if not installed: install.packages('remotes')</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"RubD/Giotto"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># compilation problems (gfortran)?</span></span>
<span><span class="co"># this version does not require C compilation</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"RubD/Giotto@cless"</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level4">
<h4 id="required-python-modules">Required python modules<a class="anchor" aria-label="anchor" href="#required-python-modules"></a>
</h4>
<p>These are necessary to run all available analyses, but can be installed automatically.</p>
<p>Required python modules:<br>
- pandas<br>
- python-igraph (igraph)<br>
- networkx<br>
- leidenalg<br>
- python-louvain (community)<br>
- smfishHmrf<br>
- python.app (!!OSX only!!)<br>
- scikit-learn</p>
<div class="section level5">
<h5 id="automatic-installation">Automatic installation<a class="anchor" aria-label="anchor" href="#automatic-installation"></a>
</h5>
<p>The python modules will be installed automatically in a miniconda environment when installing Giotto. However, it will ask you whether you want to install them and you can opt out and select your preferred python path. In that case you need to do a manual installation of the python modules.</p>
</div>
<div class="section level5">
<h5 id="manual-installation">Manual installation<a class="anchor" aria-label="anchor" href="#manual-installation"></a>
</h5>
<p>See <a href="./articles/installation_issues.html#python-manual-installation">python installation</a></p>
</div>
</div>
<div class="section level4">
<h4 id="giotto-viewer">Giotto Viewer<a class="anchor" aria-label="anchor" href="#giotto-viewer"></a>
</h4>
<p>See <a href="http://spatial.rc.fas.harvard.edu/spatialgiotto/giotto.install.native.html" class="external-link">link</a></p>
<p> </p>
</div>
</div>
<div class="section level2">
<h2 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h2>
<ul>
<li>see <a href="https://github.com/RubD/spatial-datasets" class="external-link uri">https://github.com/RubD/spatial-datasets</a> to find raw and pre-processed input data and Giotto scripts (in progress).</li>
<li>typical run time range for the different datasets on a personal computer is around 10~45 mins.<br>
</li>
<li>click on the image and try them out yourself.<br>
</li>
<li>all examples are gradually updated to the latest Giotto version [work in progress]</li>
</ul>
<p><a href="./articles/mouse_seqFISH_cortex_200914.html"><img src="./inst/images/general_figs/cortex_image_summary.png" alt="seqFISH"></a> <a href="./articles/mouse_merFISH_preoptic_region_200909.html"><img src="./inst/images/general_figs/merFISH_hypoth_image_summary.png" alt="merFISH"></a> <a href="./articles/mouse_starmap_cortex_200917.html"><img src="./inst/images/general_figs/starmap_cortex_image_summary.png" alt="STARmap"></a> <a href="./articles/mouse_visium_brain_200918.html"><img src="./inst/images/general_figs/visium_brain_image_summary.png" alt="Visium_brain"></a> <a href="./articles/mouse_visium_kidney_200916.html"><img src="./inst/images/general_figs/visium_kidney_image_summary.png" alt="Visium_kidney"></a> <a href="./articles/human_cycif_PDAC_200916.html"><img src="./inst/images/general_figs/cyCIF_PDAC_image_summary.png" alt="CyCIF"></a> <a href="./articles/mouse_osmFISH_SScortex_200915.html"><img src="./inst/images/general_figs/osmFISH_SS_cortex_image_summary.png" alt="osmFISH"></a> <a href="./articles/mouse_CODEX_spleen_200921.html"><img src="./inst/images/general_figs/CODEX_spleen_image_summary.png" alt="CODEX"></a></p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li><p><a href="https://www.biorxiv.org/content/10.1101/701680v2" class="external-link">Dries, R., Zhu, Q. et al. Giotto, a toolbox for integrative analysis and visualization of spatial expression data. bioRxiv 701680 (2019).</a> <a href="doi:10.1101/701680" class="uri">doi:10.1101/701680</a></p></li>
<li><p><a href="https://www.nature.com/articles/s41586-019-1049-y" class="external-link">Eng, C.-H. L. et al. Transcriptome-scale super-resolved imaging in tissues by RNA seqFISH+. Nature 1 (2019).</a> <a href="doi:10.1038/s41586-019-1049-y" class="uri">doi:10.1038/s41586-019-1049-y</a></p></li>
<li><p><a href="https://www.nature.com/articles/nbt.4260" class="external-link">Zhu, Q., Shah, S., Dries, R., Cai, L. &amp; Yuan, G.-C. Identification of spatially associated subpopulations by combining scRNAseq and sequential fluorescence in situ hybridization data. Nature Biotechnology (2018).</a> <a href="doi:10.1038/nbt.4260" class="uri">doi:10.1038/nbt.4260</a></p></li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/RubD/Giotto/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/RubD/Giotto/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> | file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing Giotto</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.drieslab.com/" class="external-link">Ruben Dries</a> <br><small class="roles"> Author, maintainer </small>  </li>
<li>Qian Zhu <br><small class="roles"> Author </small>  </li>
<li>Huipeng Li <br><small class="roles"> Author </small>  </li>
<li>Rui Dong <br><small class="roles"> Author </small>  </li>
<li>Guo-Cheng Yuan <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://www.drieslab.com/" class="external-link">Ruben Dries</a>, Qian Zhu, Huipeng Li, Rui Dong, Guo-Cheng Yuan.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
