<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>seqFISH+ cortex • Giotto</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="seqFISH+ cortex">
<meta property="og:description" content="Giotto">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-109153024-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-109153024-2');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Giotto</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.6.9043</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Start
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/getting_started.html#description">Description</a>
    </li>
    <li>
      <a href="../articles/getting_started.html#workflow-diagram">Workflow</a>
    </li>
    <li>
      <a href="../articles/getting_started.html#howtos">HOWTOs</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Documentation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Datasets
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/mini_seqfish.html">mini seqFISH</a>
    </li>
    <li>
      <a href="../articles/mini_visium.html">mini visium (spots)</a>
    </li>
    <li>
      <a href="../articles/mini_starmap.html">mini 3D STARmap</a>
    </li>
    <li>
      <a href="../articles/mouse_seqFISH_cortex_200914.html">mouse seqFISH cortex</a>
    </li>
    <li>
      <a href="../articles/mouse_merFISH_preoptic_region_200601.html">mouse hypoth. preopt. region</a>
    </li>
    <li>
      <a href="../articles/mouse_starmap_cortex.html">mouse STARmap cortex</a>
    </li>
    <li>
      <a href="../articles/human_cycif_PDAC_200916.html">human CyCIF PDAC</a>
    </li>
    <li>
      <a href="../articles/mouse_visium_brain_200601.html">mouse Visium brain</a>
    </li>
    <li>
      <a href="../articles/mouse_visium_kidney_200601.html">mouse Visium kidney</a>
    </li>
    <li>
      <a href="../articles/mouse_CODEX_spleen.html">mouse CODEX spleen</a>
    </li>
    <li>
      <a href="../articles/mouse_osmFISH_SScortex_200915.html">mouse osmFISH SScortex</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/faqs.html">FAQ</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/RubD/Giotto">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="mouse_seqFISH_cortex_200914_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>seqFISH+ cortex</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/RubD/Giotto/blob/master/vignettes/mouse_seqFISH_cortex_200914.Rmd"><code>vignettes/mouse_seqFISH_cortex_200914.Rmd</code></a></small>
      <div class="hidden name"><code>mouse_seqFISH_cortex_200914.Rmd</code></div>

    </div>

    
    
<div id="start-giotto" class="section level2">
<h2 class="hasAnchor">
<a href="#start-giotto" class="anchor"></a>Start Giotto</h2>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://rubd.github.io/Giotto">Giotto</a></span>)

<span class="co"># 1. set working directory</span>
<span class="kw">my_working_dir</span> <span class="op">=</span> <span class="st">'/path/to/directory/'</span>

<span class="co"># 2. set giotto python path</span>
<span class="co"># set python path to your preferred python version path</span>
<span class="co"># set python path to NULL if you want to automatically install (only the 1st time) and use the giotto miniconda environment</span>
<span class="kw">python_path</span> <span class="op">=</span> <span class="kw">NULL</span> 
<span class="co">if</span>(<span class="fu"><a href="https://rdrr.io/r/base/NULL.html">is.null</a></span>(<span class="kw">python_path</span>)) {
  <span class="fu"><a href="../reference/installGiottoEnvironment.html">installGiottoEnvironment</a></span>()
}
</pre></div>
</div>
<div id="dataset-explanation" class="section level2">
<h2 class="hasAnchor">
<a href="#dataset-explanation" class="anchor"></a>Dataset explanation</h2>
<p>Several fields - containing 100’s of cells - in the mouse cortex and subventricular zone were imaged for seqFISH+. The coordinates of the cells within each field are independent of each other, so in order to visualize and process all cells together imaging fields will be stitched together by providing x and y-offset values specific to each field. These offset values are known or estimates based on the original raw image:<br><img src="../inst/images/general_figs/cortex_svz_location_fields.png" style="width:50.0%"></p>
<ul>
<li>load somatosensory (SS) cortex and subventricular zone (SVZ) gene expression matrix<br>
</li>
<li>load cell coordinates and field of view (FOV) information in order to stitch imaging fields</li>
</ul>
</div>
<div id="dataset-download" class="section level2">
<h2 class="hasAnchor">
<a href="#dataset-download" class="anchor"></a>Dataset download</h2>
<p>The seqFISH+ data to run this tutorial can be downloaded directly using the <strong>getSpatialDataset</strong> function or can also be found <a href="https://github.com/RubD/spatial-datasets/tree/master/data/2019_seqfish_plus_SScortex">here</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co"># download data to working directory ####</span>
<span class="co"># if wget is installed, set method = 'wget'</span>
<span class="fu"><a href="../reference/getSpatialDataset.html">getSpatialDataset</a></span>(dataset = <span class="st">'seqfish_SS_cortex'</span>, directory = <span class="kw">my_working_dir</span>, method = <span class="st">'wget'</span>)
</pre></div>
</div>
<div id="part-1-giotto-global-instructions-and-preparations" class="section level2">
<h2 class="hasAnchor">
<a href="#part-1-giotto-global-instructions-and-preparations" class="anchor"></a>Part 1: Giotto global instructions and preparations</h2>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="co"># 1. (optional) set Giotto instructions</span>
<span class="kw">instrs</span> <span class="op">=</span> <span class="fu"><a href="../reference/createGiottoInstructions.html">createGiottoInstructions</a></span>(save_plot = <span class="fl">TRUE</span>, 
                                  show_plot = <span class="fl">FALSE</span>,
                                  save_dir = <span class="kw">my_working_dir</span>, 
                                  python_path = <span class="kw">python_path</span>)

<span class="co"># 2. create giotto object from provided paths ####</span>
<span class="kw">expr_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="kw">my_working_dir</span>, <span class="st">"cortex_svz_expression.txt"</span>)
<span class="kw">loc_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="kw">my_working_dir</span>, <span class="st">"cortex_svz_centroids_coord.txt"</span>)
<span class="kw">meta_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="kw">my_working_dir</span>, <span class="st">"cortex_svz_centroids_annot.txt"</span>)


<span class="co"># 3. This dataset contains multiple field of views which need to be stitched together</span>

<span class="co">## first merge location and additional metadata</span>
<span class="kw">SS_locations</span> <span class="op">=</span> <span class="kw">data.table</span>::<span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span>(<span class="kw">loc_path</span>)
<span class="kw">cortex_fields</span> <span class="op">=</span> <span class="kw">data.table</span>::<span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span>(<span class="kw">meta_path</span>)
<span class="kw">SS_loc_annot</span> <span class="op">=</span> <span class="kw">data.table</span>::<span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/merge.html">merge.data.table</a></span>(<span class="kw">SS_locations</span>, <span class="kw">cortex_fields</span>, by = <span class="st">'ID'</span>)
<span class="kw">SS_loc_annot</span>[, <span class="kw">ID</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="kw">ID</span>, levels = <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">'cell_'</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">913</span>))]
<span class="kw">data.table</span>::<span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setorder.html">setorder</a></span>(<span class="kw">SS_loc_annot</span>, <span class="kw">ID</span>)

<span class="co">## create file with offset information</span>
<span class="kw">my_offset_file</span> <span class="op">=</span> <span class="kw">data.table</span>::<span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html">data.table</a></span>(field = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">6</span>),
                            x_offset = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">1654.97</span>, <span class="fl">1750.75</span>, <span class="fl">1674.35</span>, <span class="fl">675.5</span>, <span class="fl">2048</span>, <span class="fl">675</span>),
                            y_offset = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="op">-</span><span class="fl">1438.02</span>, <span class="op">-</span><span class="fl">1438.02</span>, <span class="fl">0</span>))

<span class="co">## create a stitch file</span>
<span class="kw">stitch_file</span> <span class="op">=</span> <span class="fu"><a href="../reference/stitchFieldCoordinates.html">stitchFieldCoordinates</a></span>(location_file = <span class="kw">SS_loc_annot</span>,
                                     offset_file = <span class="kw">my_offset_file</span>,
                                     cumulate_offset_x = <span class="kw">T</span>,
                                     cumulate_offset_y = <span class="kw">F</span>,
                                     field_col = <span class="st">'FOV'</span>,
                                     reverse_final_x = <span class="kw">F</span>,
                                     reverse_final_y = <span class="kw">T</span>)
<span class="kw">stitch_file</span>    <span class="op">=</span> <span class="kw">stitch_file</span>[,<span class="fu">.</span>(<span class="kw">ID</span>, <span class="kw">X_final</span>, <span class="kw">Y_final</span>)]
<span class="kw">my_offset_file</span> <span class="op">=</span> <span class="kw">my_offset_file</span>[,<span class="fu">.</span>(<span class="kw">field</span>, <span class="kw">x_offset_final</span>, <span class="kw">y_offset_final</span>)]
</pre></div>
</div>
<div id="part-2-create-giotto-object-process-data" class="section level2">
<h2 class="hasAnchor">
<a href="#part-2-create-giotto-object-process-data" class="anchor"></a>part 2: Create Giotto object &amp; process data</h2>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="co">## create Giotto object</span>
<span class="kw">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createGiottoObject.html">createGiottoObject</a></span>(raw_exprs = <span class="kw">expr_path</span>,
                                 spatial_locs = <span class="kw">stitch_file</span>,
                                 offset_file = <span class="kw">my_offset_file</span>,
                                 instructions = <span class="kw">instrs</span>)

<span class="co">## add additional annotation if wanted</span>
<span class="kw">SS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="../reference/addCellMetadata.html">addCellMetadata</a></span>(<span class="kw">SS_seqfish</span>,
                             new_metadata = <span class="kw">cortex_fields</span>,
                             by_column = <span class="kw">T</span>,
                             column_cell_ID = <span class="st">'ID'</span>)

<span class="co">## subset data to the cortex field of views</span>
<span class="kw">cell_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/pDataDT.html">pDataDT</a></span>(<span class="kw">SS_seqfish</span>)
<span class="kw">cortex_cell_ids</span> <span class="op">=</span> <span class="kw">cell_metadata</span>[<span class="kw">FOV</span> <span class="op">%in%</span> <span class="fl">0</span><span class="op">:</span><span class="fl">4</span>]<span class="op">$</span><span class="kw">cell_ID</span>
<span class="kw">SS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="../reference/subsetGiotto.html">subsetGiotto</a></span>(<span class="kw">SS_seqfish</span>, cell_ids = <span class="kw">cortex_cell_ids</span>)

<span class="co">## filter</span>
<span class="kw">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterGiotto.html">filterGiotto</a></span>(gobject = <span class="kw">SS_seqfish</span>,
                        expression_threshold = <span class="fl">1</span>,
                        gene_det_in_min_cells = <span class="fl">10</span>,
                        min_det_genes_per_cell = <span class="fl">10</span>,
                        expression_values = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'raw'</span>),
                        verbose = <span class="kw">T</span>)

<span class="co">## normalize</span>
<span class="kw">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span>(gobject = <span class="kw">SS_seqfish</span>, scalefactor = <span class="fl">6000</span>, verbose = <span class="kw">T</span>)

<span class="co">## add gene &amp; cell statistics</span>
<span class="kw">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addStatistics.html">addStatistics</a></span>(gobject = <span class="kw">SS_seqfish</span>)

<span class="co">## adjust expression matrix for technical or known variables</span>
<span class="kw">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adjustGiottoMatrix.html">adjustGiottoMatrix</a></span>(gobject = <span class="kw">SS_seqfish</span>, expression_values = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'normalized'</span>),
                              batch_columns = <span class="kw">NULL</span>, covariate_columns = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'nr_genes'</span>, <span class="st">'total_expr'</span>),
                              return_gobject = <span class="fl">TRUE</span>,
                              update_slot = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'custom'</span>))

<span class="co">## visualize</span>
<span class="fu"><a href="../reference/spatPlot.html">spatPlot</a></span>(gobject = <span class="kw">SS_seqfish</span>,
         save_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(save_name = <span class="st">'2_a_spatplot'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/2_a_spatplot.png" style="width:50.0%"></p>
</div>
<div id="part-3-dimension-reduction" class="section level2">
<h2 class="hasAnchor">
<a href="#part-3-dimension-reduction" class="anchor"></a>part 3: dimension reduction</h2>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co">## highly variable genes (HVG)</span>
<span class="kw">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHVG.html">calculateHVG</a></span>(gobject = <span class="kw">SS_seqfish</span>, method = <span class="st">'cov_loess'</span>, difference_in_cov = <span class="fl">0.1</span>,
                           save_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(save_name = <span class="st">'3_a_HVGplot'</span>, base_height = <span class="fl">5</span>, base_width = <span class="fl">5</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/3_a_HVGplot.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="co">## select genes based on HVG and gene statistics, both found in gene metadata</span>
<span class="kw">gene_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/fDataDT.html">fDataDT</a></span>(<span class="kw">SS_seqfish</span>)
<span class="kw">featgenes</span> <span class="op">=</span> <span class="kw">gene_metadata</span>[<span class="kw">hvg</span> <span class="op">==</span> <span class="st">'yes'</span> <span class="op">&amp;</span> <span class="kw">perc_cells</span> <span class="op">&gt;</span> <span class="fl">4</span> <span class="op">&amp;</span> <span class="kw">mean_expr_det</span> <span class="op">&gt;</span> <span class="fl">0.5</span>]<span class="op">$</span><span class="kw">gene_ID</span>

<span class="co">## run PCA on expression values (default)</span>
<span class="kw">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runPCA.html">runPCA</a></span>(gobject = <span class="kw">SS_seqfish</span>, genes_to_use = <span class="kw">featgenes</span>, scale_unit = <span class="kw">F</span>, center = <span class="kw">F</span>)
<span class="fu"><a href="../reference/screePlot.html">screePlot</a></span>(<span class="kw">SS_seqfish</span>, save_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(save_name = <span class="st">'3_b_screeplot'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/3_b_screeplot.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="fu"><a href="../reference/plotPCA.html">plotPCA</a></span>(gobject = <span class="kw">SS_seqfish</span>,
        save_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(save_name = <span class="st">'3_c_PCA_reduction'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/3_c_PCA_reduction.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="co">## run UMAP and tSNE on PCA space (default)</span>
<span class="kw">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span>(<span class="kw">SS_seqfish</span>, dimensions_to_use = <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, n_threads = <span class="fl">10</span>)
<span class="fu"><a href="../reference/plotUMAP.html">plotUMAP</a></span>(gobject = <span class="kw">SS_seqfish</span>,
         save_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(save_name = <span class="st">'3_d_UMAP_reduction'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/3_d_UMAP_reduction.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="kw">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runtSNE.html">runtSNE</a></span>(<span class="kw">SS_seqfish</span>, dimensions_to_use = <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>)
<span class="fu"><a href="../reference/plotTSNE.html">plotTSNE</a></span>(gobject = <span class="kw">SS_seqfish</span>,
         save_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(save_name = <span class="st">'3_e_tSNE_reduction'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/3_e_tSNE_reduction.png" style="width:50.0%"></p>
</div>
<div id="part-4-cluster" class="section level2">
<h2 class="hasAnchor">
<a href="#part-4-cluster" class="anchor"></a>part 4: cluster</h2>
<div class="sourceCode" id="cb10"><pre class="downlit">

<span class="co">## sNN network (default)</span>
<span class="kw">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createNearestNetwork.html">createNearestNetwork</a></span>(gobject = <span class="kw">SS_seqfish</span>, dimensions_to_use = <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, k = <span class="fl">15</span>)
<span class="co">## Leiden clustering</span>
<span class="kw">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span>(gobject = <span class="kw">SS_seqfish</span>, resolution = <span class="fl">0.4</span>, n_iterations = <span class="fl">1000</span>)
<span class="fu"><a href="../reference/plotUMAP.html">plotUMAP</a></span>(gobject = <span class="kw">SS_seqfish</span>,
         cell_color = <span class="st">'leiden_clus'</span>, show_NN_network = <span class="kw">T</span>, point_size = <span class="fl">2.5</span>,
         save_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(save_name = <span class="st">'4_a_UMAP_leiden'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/4_a_UMAP_leiden.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="co">## Leiden subclustering for specified clusters</span>
<span class="kw">SS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="../reference/doLeidenSubCluster.html">doLeidenSubCluster</a></span>(gobject = <span class="kw">SS_seqfish</span>, cluster_column = <span class="st">'leiden_clus'</span>,
                             resolution = <span class="fl">0.2</span>, k_neighbors = <span class="fl">10</span>,
                             hvg_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(method = <span class="st">'cov_loess'</span>, difference_in_cov = <span class="fl">0.1</span>),
                             pca_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(expression_values = <span class="st">'normalized'</span>, scale_unit = <span class="kw">F</span>),
                             nn_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(dimensions_to_use = <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>),
                             selected_clusters = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">7</span>),
                             name = <span class="st">'sub_leiden_clus_select'</span>)

<span class="co">## set colors for clusters</span>
<span class="kw">subleiden_order</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>( <span class="fl">1.1</span>, <span class="fl">5.1</span>, <span class="fl">5.2</span>,  <span class="fl">2.1</span>, <span class="fl">3.1</span>,
                     <span class="fl">4.1</span>, <span class="fl">6.2</span>, <span class="fl">6.1</span>,
                     <span class="fl">7.1</span>,  <span class="fl">7.2</span>, <span class="fl">9.1</span>, <span class="fl">8.1</span>)
<span class="kw">subleiden_colors</span> <span class="op">=</span> <span class="kw">Giotto</span>:::<span class="fu"><a href="../reference/getDistinctColors.html">getDistinctColors</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="kw">subleiden_order</span>)) 
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw">subleiden_colors</span>) <span class="op">=</span> <span class="kw">subleiden_order</span>

<span class="fu"><a href="../reference/plotUMAP.html">plotUMAP</a></span>(gobject = <span class="kw">SS_seqfish</span>,
         cell_color = <span class="st">'sub_leiden_clus_select'</span>, cell_color_code = <span class="kw">subleiden_colors</span>,
         show_NN_network = <span class="kw">T</span>, point_size = <span class="fl">2.5</span>, show_center_label = <span class="kw">F</span>, 
         legend_text = <span class="fl">12</span>, legend_symbol_size = <span class="fl">3</span>,
         save_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(save_name = <span class="st">'4_b_UMAP_leiden_subcluster'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/4_b_UMAP_leiden_subcluster.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="co">## show cluster relationships</span>
<span class="fu"><a href="../reference/showClusterHeatmap.html">showClusterHeatmap</a></span>(gobject = <span class="kw">SS_seqfish</span>, cluster_column = <span class="st">'sub_leiden_clus_select'</span>,
                   save_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(save_name = <span class="st">'4_c_heatmap'</span>, units = <span class="st">'cm'</span>),
                   row_names_gp = <span class="kw">grid</span>::<span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span>(fontsize = <span class="fl">9</span>), column_names_gp = <span class="kw">grid</span>::<span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span>(fontsize = <span class="fl">9</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/4_c_heatmap.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="fu"><a href="../reference/showClusterDendrogram.html">showClusterDendrogram</a></span>(<span class="kw">SS_seqfish</span>, h = <span class="fl">0.5</span>, rotate = <span class="kw">T</span>, cluster_column = <span class="st">'sub_leiden_clus_select'</span>,
                      save_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(save_name = <span class="st">'4_d_dendro'</span>, units = <span class="st">'cm'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/4_d_dendro.png" style="width:50.0%"></p>
</div>
<div id="part-5-visualize-spatial-and-expression-space" class="section level2">
<h2 class="hasAnchor">
<a href="#part-5-visualize-spatial-and-expression-space" class="anchor"></a>part 5: visualize spatial and expression space</h2>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="co"># expression and spatial</span>
<span class="fu"><a href="../reference/spatDimPlot.html">spatDimPlot</a></span>(gobject = <span class="kw">SS_seqfish</span>, cell_color = <span class="st">'sub_leiden_clus_select'</span>, 
            cell_color_code = <span class="kw">subleiden_colors</span>,
            dim_point_size = <span class="fl">2</span>, spat_point_size = <span class="fl">2</span>,
            save_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(save_name = <span class="st">'5_a_covis_leiden'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/5_a_covis_leiden.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="co"># selected groups and provide new colors</span>
<span class="kw">groups_of_interest</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">6.1</span>, <span class="fl">6.2</span>, <span class="fl">7.1</span>, <span class="fl">7.2</span>)
<span class="kw">group_colors</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'red'</span>, <span class="st">'green'</span>, <span class="st">'blue'</span>, <span class="st">'purple'</span>); <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw">group_colors</span>) <span class="op">=</span> <span class="kw">groups_of_interest</span>

<span class="fu"><a href="../reference/spatDimPlot.html">spatDimPlot</a></span>(gobject = <span class="kw">SS_seqfish</span>, cell_color = <span class="st">'sub_leiden_clus_select'</span>, 
            dim_point_size = <span class="fl">2</span>, spat_point_size = <span class="fl">2</span>,
            select_cell_groups = <span class="kw">groups_of_interest</span>, cell_color_code = <span class="kw">group_colors</span>,
            save_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(save_name = <span class="st">'5_b_covis_leiden_selected'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/5_b_covis_leiden_selected.png" style="width:50.0%"></p>
</div>
<div id="part-6-cell-type-marker-gene-detection" class="section level2">
<h2 class="hasAnchor">
<a href="#part-6-cell-type-marker-gene-detection" class="anchor"></a>part 6: cell type marker gene detection</h2>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="co">## gini ##</span>
<span class="kw">gini_markers_subclusters</span> <span class="op">=</span> <span class="fu"><a href="../reference/findMarkers_one_vs_all.html">findMarkers_one_vs_all</a></span>(gobject = <span class="kw">SS_seqfish</span>,
                                                  method = <span class="st">'gini'</span>,
                                                  expression_values = <span class="st">'normalized'</span>,
                                                  cluster_column = <span class="st">'sub_leiden_clus_select'</span>,
                                                  min_genes = <span class="fl">20</span>,
                                                  min_expr_gini_score = <span class="fl">0.5</span>,
                                                  min_det_gini_score = <span class="fl">0.5</span>)
<span class="kw">topgenes_gini</span> <span class="op">=</span> <span class="kw">gini_markers_subclusters</span>[, <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">.SD</span>, <span class="fl">2</span>), by = <span class="st">'cluster'</span>]

<span class="co"># violinplot</span>
<span class="fu"><a href="../reference/violinPlot.html">violinPlot</a></span>(<span class="kw">SS_seqfish</span>, genes = <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="kw">topgenes_gini</span><span class="op">$</span><span class="kw">genes</span>), cluster_column = <span class="st">'sub_leiden_clus_select'</span>,
           strip_text = <span class="fl">8</span>, strip_position = <span class="st">'right'</span>, cluster_custom_order = <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="kw">topgenes_gini</span><span class="op">$</span><span class="kw">cluster</span>),
           save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'6_a_violinplot_gini'</span>, base_width = <span class="fl">5</span>, base_height = <span class="fl">10</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/6_a_violinplot_gini.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="co"># cluster heatmap</span>
<span class="kw">topgenes_gini2</span> <span class="op">=</span> <span class="kw">gini_markers_subclusters</span>[, <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">.SD</span>, <span class="fl">6</span>), by = <span class="st">'cluster'</span>]
<span class="fu"><a href="../reference/plotMetaDataHeatmap.html">plotMetaDataHeatmap</a></span>(<span class="kw">SS_seqfish</span>, selected_genes = <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="kw">topgenes_gini2</span><span class="op">$</span><span class="kw">genes</span>), 
                    custom_gene_order = <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="kw">topgenes_gini2</span><span class="op">$</span><span class="kw">genes</span>),
                    custom_cluster_order = <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="kw">topgenes_gini2</span><span class="op">$</span><span class="kw">cluster</span>),
                    metadata_cols = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'sub_leiden_clus_select'</span>), x_text_size = <span class="fl">10</span>, y_text_size = <span class="fl">10</span>,
                    save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'6_b_metaheatmap_gini'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/6_b_metaheatmap_gini.png" style="width:50.0%"></p>
</div>
<div id="part-7-cell-type-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#part-7-cell-type-annotation" class="anchor"></a>part 7: cell type annotation</h2>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="co">## general cell types</span>
<span class="co"># create vector with names</span>
<span class="kw">clusters_cell_types_cortex</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'L6 eNeuron'</span>, <span class="st">'L4 eNeuron'</span>, <span class="st">'L2/3 eNeuron'</span>, <span class="st">'L5 eNeuron'</span>, 
                               <span class="st">'Lhx6 iNeuron'</span>, <span class="st">'Adarb2 iNeuron'</span>, 
                               <span class="st">'endothelial'</span>, <span class="st">'mural'</span>,
                               <span class="st">'OPC'</span>,<span class="st">'Olig'</span>,
                               <span class="st">'astrocytes'</span>, <span class="st">'microglia'</span>)
                              
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw">clusters_cell_types_cortex</span>) <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1.1</span>, <span class="fl">2.1</span>, <span class="fl">3.1</span>, <span class="fl">4.1</span>,
                                      <span class="fl">5.1</span>, <span class="fl">5.2</span>,
                                      <span class="fl">6.1</span>, <span class="fl">6.2</span>, 
                                      <span class="fl">7.1</span>, <span class="fl">7.2</span>,
                                      <span class="fl">8.1</span>, <span class="fl">9.1</span>)

<span class="kw">SS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="../reference/annotateGiotto.html">annotateGiotto</a></span>(gobject = <span class="kw">SS_seqfish</span>, annotation_vector = <span class="kw">clusters_cell_types_cortex</span>,
                         cluster_column = <span class="st">'sub_leiden_clus_select'</span>, name = <span class="st">'cell_types'</span>)

<span class="co"># cell type order and colors</span>
<span class="kw">cell_type_order</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'L6 eNeuron'</span>, <span class="st">'L5 eNeuron'</span>, <span class="st">'L4 eNeuron'</span>, <span class="st">'L2/3 eNeuron'</span>,
                    <span class="st">'astrocytes'</span>, <span class="st">'Olig'</span>, <span class="st">'OPC'</span>,<span class="st">'Adarb2 iNeuron'</span>, <span class="st">'Lhx6 iNeuron'</span>,
                    <span class="st">'endothelial'</span>, <span class="st">'mural'</span>, <span class="st">'microglia'</span>)
<span class="kw">cell_type_colors</span> <span class="op">=</span> <span class="kw">Giotto</span>:::<span class="fu"><a href="../reference/getDistinctColors.html">getDistinctColors</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="kw">cell_type_order</span>)) 
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw">cell_type_colors</span>) <span class="op">=</span> <span class="kw">cell_type_order</span>

<span class="co">## violinplot</span>
<span class="fu"><a href="../reference/violinPlot.html">violinPlot</a></span>(gobject = <span class="kw">SS_seqfish</span>, genes = <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="kw">topgenes_gini</span><span class="op">$</span><span class="kw">genes</span>),
           strip_text = <span class="fl">7</span>, strip_position = <span class="st">'right'</span>, 
           cluster_custom_order = <span class="kw">cell_type_order</span>,
           cluster_column = <span class="st">'cell_types'</span>, color_violin = <span class="st">'cluster'</span>,
           save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'7_a_violinplot'</span>, base_width = <span class="fl">5</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/7_a_violinplot.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="co">## co-visualization</span>
<span class="fu"><a href="../reference/spatDimPlot.html">spatDimPlot</a></span>(gobject = <span class="kw">SS_seqfish</span>, cell_color = <span class="st">'cell_types'</span>,
            dim_point_size = <span class="fl">2</span>, spat_point_size = <span class="fl">2</span>, dim_show_cluster_center = <span class="kw">F</span>, dim_show_center_label = <span class="kw">T</span>,
            save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'7_b_covisualization'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/7_b_covisualization.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="co">## heatmap genes vs cells</span>
<span class="kw">gini_markers_subclusters</span>[, <span class="kw">cell_types</span> <span class="op">:=</span> <span class="kw">clusters_cell_types_cortex</span>[<span class="kw">cluster</span>] ]
<span class="kw">gini_markers_subclusters</span>[, <span class="kw">cell_types</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="kw">cell_types</span>, <span class="kw">cell_type_order</span>)]
<span class="kw">data.table</span>::<span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setorder.html">setorder</a></span>(<span class="kw">gini_markers_subclusters</span>, <span class="kw">cell_types</span>)

<span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span>(gobject = <span class="kw">SS_seqfish</span>,
            genes = <span class="kw">gini_markers_subclusters</span>[, <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">.SD</span>, <span class="fl">3</span>), by = <span class="st">'cell_types'</span>]<span class="op">$</span><span class="kw">genes</span>, 
            gene_order = <span class="st">'custom'</span>,
            gene_custom_order = <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="kw">gini_markers_subclusters</span>[, <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">.SD</span>, <span class="fl">3</span>), by = <span class="st">'cluster'</span>]<span class="op">$</span><span class="kw">genes</span>),
            cluster_column = <span class="st">'cell_types'</span>, cluster_order = <span class="st">'custom'</span>,
            cluster_custom_order = <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="kw">gini_markers_subclusters</span>[, <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">.SD</span>, <span class="fl">3</span>), by = <span class="st">'cell_types'</span>]<span class="op">$</span><span class="kw">cell_types</span>), 
            legend_nrows = <span class="fl">2</span>,
            save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'7_c_heatmap'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/7_c_heatmap.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb21"><pre class="downlit">
<span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span>(gobject = <span class="kw">SS_seqfish</span>,
            cluster_color_code = <span class="kw">cell_type_colors</span>,
            genes = <span class="kw">gini_markers_subclusters</span>[, <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">.SD</span>, <span class="fl">6</span>), by = <span class="st">'cell_types'</span>]<span class="op">$</span><span class="kw">genes</span>,
            gene_order = <span class="st">'custom'</span>,
            gene_label_selection = <span class="kw">gini_markers_subclusters</span>[, <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">.SD</span>, <span class="fl">2</span>), by = <span class="st">'cluster'</span>]<span class="op">$</span><span class="kw">genes</span>,
            gene_custom_order = <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="kw">gini_markers_subclusters</span>[, <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">.SD</span>, <span class="fl">6</span>), by = <span class="st">'cluster'</span>]<span class="op">$</span><span class="kw">genes</span>),
            cluster_column = <span class="st">'cell_types'</span>, cluster_order = <span class="st">'custom'</span>,
            cluster_custom_order = <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="kw">gini_markers_subclusters</span>[, <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">.SD</span>, <span class="fl">3</span>), by = <span class="st">'cell_types'</span>]<span class="op">$</span><span class="kw">cell_types</span>), 
            legend_nrows = <span class="fl">2</span>,
            save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'7_d_heatmap_selected'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/7_d_heatmap_selected.png" style="width:50.0%"></p>
</div>
<div id="part-8-spatial-grid" class="section level2">
<h2 class="hasAnchor">
<a href="#part-8-spatial-grid" class="anchor"></a>part 8: spatial grid</h2>
<div class="sourceCode" id="cb22"><pre class="downlit">
<span class="co">## spatial grid</span>
<span class="kw">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSpatialGrid.html">createSpatialGrid</a></span>(gobject = <span class="kw">SS_seqfish</span>,
                             sdimx_stepsize = <span class="fl">500</span>,
                             sdimy_stepsize = <span class="fl">500</span>,
                             minimum_padding = <span class="fl">50</span>)

<span class="fu"><a href="../reference/spatPlot.html">spatPlot</a></span>(gobject = <span class="kw">SS_seqfish</span>, show_grid = <span class="kw">T</span>, point_size = <span class="fl">1.5</span>,
         save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'8_a_grid'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/8_a_grid.png" style="width:50.0%"></p>
</div>
<div id="part-9-spatial-network" class="section level2">
<h2 class="hasAnchor">
<a href="#part-9-spatial-network" class="anchor"></a>part 9: spatial network</h2>
<div class="sourceCode" id="cb23"><pre class="downlit">

<span class="co">## delaunay network: stats + creation</span>
<span class="fu"><a href="../reference/plotStatDelaunayNetwork.html">plotStatDelaunayNetwork</a></span>(gobject = <span class="kw">SS_seqfish</span>, maximum_distance = <span class="fl">400</span>, save_plot = <span class="kw">F</span>)
<span class="kw">SS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="../reference/createSpatialNetwork.html">createSpatialNetwork</a></span>(gobject = <span class="kw">SS_seqfish</span>, minimum_k = <span class="fl">2</span>, maximum_distance_delaunay = <span class="fl">400</span>)

<span class="co">## create spatial networks based on k and/or distance from centroid</span>
<span class="kw">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSpatialNetwork.html">createSpatialNetwork</a></span>(gobject = <span class="kw">SS_seqfish</span>, method = <span class="st">'kNN'</span>, k = <span class="fl">5</span>, name = <span class="st">'spatial_network'</span>)
<span class="kw">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSpatialNetwork.html">createSpatialNetwork</a></span>(gobject = <span class="kw">SS_seqfish</span>, method = <span class="st">'kNN'</span>, k = <span class="fl">10</span>, name = <span class="st">'large_network'</span>)
<span class="kw">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSpatialNetwork.html">createSpatialNetwork</a></span>(gobject = <span class="kw">SS_seqfish</span>, method = <span class="st">'kNN'</span>, k = <span class="fl">100</span>,
                                maximum_distance_knn = <span class="fl">200</span>, minimum_k = <span class="fl">2</span>, name = <span class="st">'distance_network'</span>)

<span class="co">## visualize different spatial networks on first field (~ layer 1)</span>
<span class="kw">cell_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/pDataDT.html">pDataDT</a></span>(<span class="kw">SS_seqfish</span>)
<span class="kw">field1_ids</span> <span class="op">=</span> <span class="kw">cell_metadata</span>[<span class="kw">FOV</span> <span class="op">==</span> <span class="fl">0</span>]<span class="op">$</span><span class="kw">cell_ID</span>
<span class="kw">subSS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="../reference/subsetGiotto.html">subsetGiotto</a></span>(<span class="kw">SS_seqfish</span>, cell_ids = <span class="kw">field1_ids</span>)

<span class="fu"><a href="../reference/spatPlot.html">spatPlot</a></span>(gobject = <span class="kw">subSS_seqfish</span>, show_network = <span class="kw">T</span>,
         network_color = <span class="st">'blue'</span>, spatial_network_name = <span class="st">'Delaunay_network'</span>,
         point_size = <span class="fl">2.5</span>, cell_color = <span class="st">'cell_types'</span>, 
         save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'9_a_spatial_network_delaunay'</span>, base_height = <span class="fl">6</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/9_a_spatial_network_delaunay.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb24"><pre class="downlit">
<span class="fu"><a href="../reference/spatPlot.html">spatPlot</a></span>(gobject = <span class="kw">subSS_seqfish</span>, show_network = <span class="kw">T</span>,
         network_color = <span class="st">'blue'</span>, spatial_network_name = <span class="st">'spatial_network'</span>,
         point_size = <span class="fl">2.5</span>, cell_color = <span class="st">'cell_types'</span>,
         save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'9_b_spatial_network_k3'</span>, base_height = <span class="fl">6</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/9_b_spatial_network_k3.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb25"><pre class="downlit">
<span class="fu"><a href="../reference/spatPlot.html">spatPlot</a></span>(gobject = <span class="kw">subSS_seqfish</span>, show_network = <span class="kw">T</span>,
         network_color = <span class="st">'blue'</span>, spatial_network_name = <span class="st">'large_network'</span>,
         point_size = <span class="fl">2.5</span>, cell_color = <span class="st">'cell_types'</span>,
         save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'9_c_spatial_network_k10'</span>, base_height = <span class="fl">6</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/9_c_spatial_network_k10.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb26"><pre class="downlit">
<span class="fu"><a href="../reference/spatPlot.html">spatPlot</a></span>(gobject = <span class="kw">subSS_seqfish</span>, show_network = <span class="kw">T</span>,
         network_color = <span class="st">'blue'</span>, spatial_network_name = <span class="st">'distance_network'</span>,
         point_size = <span class="fl">2.5</span>, cell_color = <span class="st">'cell_types'</span>,
         save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'9_d_spatial_network_dist'</span>, base_height = <span class="fl">6</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/9_d_spatial_network_dist.png" style="width:50.0%"></p>
</div>
<div id="part-10-spatial-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#part-10-spatial-genes" class="anchor"></a>part 10: spatial genes</h2>
<div id="individual-spatial-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#individual-spatial-genes" class="anchor"></a>Individual spatial genes</h3>
<div class="sourceCode" id="cb27"><pre class="downlit">
<span class="co"># 3 new methods to identify spatial genes</span>
<span class="kw">km_spatialgenes</span> <span class="op">=</span> <span class="fu"><a href="../reference/binSpect.html">binSpect</a></span>(<span class="kw">SS_seqfish</span>)

<span class="fu"><a href="../reference/spatGenePlot.html">spatGenePlot</a></span>(<span class="kw">SS_seqfish</span>, expression_values = <span class="st">'scaled'</span>, genes = <span class="kw">km_spatialgenes</span>[<span class="fl">1</span><span class="op">:</span><span class="fl">4</span>]<span class="op">$</span><span class="kw">genes</span>,
             point_shape = <span class="st">'border'</span>, point_border_stroke = <span class="fl">0.1</span>,
             show_network = <span class="kw">F</span>, network_color = <span class="st">'lightgrey'</span>, point_size = <span class="fl">2.5</span>, 
             cow_n_col = <span class="fl">2</span>,
             save_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(save_name = <span class="st">'10_a_spatialgenes_km'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/10_a_spatialgenes_km.png" style="width:50.0%"></p>
</div>
<div id="spatial-genes-co-expression-modules" class="section level3">
<h3 class="hasAnchor">
<a href="#spatial-genes-co-expression-modules" class="anchor"></a>Spatial genes co-expression modules</h3>
<div class="sourceCode" id="cb28"><pre class="downlit">
<span class="co">## spatial co-expression patterns ##</span>
<span class="kw">ext_spatial_genes</span> <span class="op">=</span> <span class="kw">km_spatialgenes</span>[<span class="fl">1</span><span class="op">:</span><span class="fl">500</span>]<span class="op">$</span><span class="kw">genes</span>

<span class="co"># 1. calculate gene spatial correlation and single-cell correlation </span>
<span class="co"># create spatial correlation object</span>
<span class="kw">spat_cor_netw_DT</span> <span class="op">=</span> <span class="fu"><a href="../reference/detectSpatialCorGenes.html">detectSpatialCorGenes</a></span>(<span class="kw">SS_seqfish</span>, 
                                         method = <span class="st">'network'</span>, spatial_network_name = <span class="st">'Delaunay_network'</span>,
                                         subset_genes = <span class="kw">ext_spatial_genes</span>)


<span class="co"># 2. cluster correlated genes &amp; visualize</span>
<span class="kw">spat_cor_netw_DT</span> <span class="op">=</span> <span class="fu"><a href="../reference/clusterSpatialCorGenes.html">clusterSpatialCorGenes</a></span>(<span class="kw">spat_cor_netw_DT</span>, name = <span class="st">'spat_netw_clus'</span>, k = <span class="fl">8</span>)

<span class="fu"><a href="../reference/heatmSpatialCorGenes.html">heatmSpatialCorGenes</a></span>(<span class="kw">SS_seqfish</span>, spatCorObject = <span class="kw">spat_cor_netw_DT</span>, use_clus_name = <span class="st">'spat_netw_clus'</span>,
                     save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'10_b_spatialcoexpression_heatmap'</span>,
                                    base_height = <span class="fl">6</span>, base_width = <span class="fl">8</span>, units = <span class="st">'cm'</span>), 
                     heatmap_legend_param = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(title = <span class="kw">NULL</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/10_b_spatialcoexpression_heatmap.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb29"><pre class="downlit">
<span class="co"># 3. rank spatial correlated clusters and show genes for selected clusters</span>
<span class="kw">netw_ranks</span> <span class="op">=</span> <span class="fu"><a href="../reference/rankSpatialCorGroups.html">rankSpatialCorGroups</a></span>(<span class="kw">SS_seqfish</span>, spatCorObject = <span class="kw">spat_cor_netw_DT</span>, use_clus_name = <span class="st">'spat_netw_clus'</span>,
                                  save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'10_c_spatialcoexpression_rank'</span>,
                                                 base_height = <span class="fl">3</span>, base_width = <span class="fl">5</span>))

<span class="kw">top_netw_spat_cluster</span> <span class="op">=</span> <span class="fu"><a href="../reference/showSpatialCorGenes.html">showSpatialCorGenes</a></span>(<span class="kw">spat_cor_netw_DT</span>, use_clus_name = <span class="st">'spat_netw_clus'</span>,
                                            selected_clusters = <span class="fl">6</span>, show_top_genes = <span class="fl">1</span>)
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/10_c_spatialcoexpression_rank.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb30"><pre class="downlit">
<span class="co"># 4. create metagene enrichment score for clusters</span>
<span class="kw">cluster_genes_DT</span> <span class="op">=</span> <span class="fu"><a href="../reference/showSpatialCorGenes.html">showSpatialCorGenes</a></span>(<span class="kw">spat_cor_netw_DT</span>, use_clus_name = <span class="st">'spat_netw_clus'</span>, show_top_genes = <span class="fl">1</span>)
<span class="kw">cluster_genes</span> <span class="op">=</span> <span class="kw">cluster_genes_DT</span><span class="op">$</span><span class="kw">clus</span>; <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw">cluster_genes</span>) <span class="op">=</span> <span class="kw">cluster_genes_DT</span><span class="op">$</span><span class="kw">gene_ID</span>

<span class="kw">SS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="../reference/createMetagenes.html">createMetagenes</a></span>(<span class="kw">SS_seqfish</span>, gene_clusters = <span class="kw">cluster_genes</span>, name = <span class="st">'cluster_metagene'</span>)

<span class="fu"><a href="../reference/spatCellPlot.html">spatCellPlot</a></span>(<span class="kw">SS_seqfish</span>,
             spat_enr_names = <span class="st">'cluster_metagene'</span>,
             cell_annotation_values = <span class="kw">netw_ranks</span><span class="op">$</span><span class="kw">clusters</span>,
             point_size = <span class="fl">1.5</span>, cow_n_col = <span class="fl">3</span>,
             save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'10_d_spatialcoexpression_metagenes'</span>,
                            base_width = <span class="fl">11</span>, base_height = <span class="fl">6</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/10_d_spatialcoexpression_metagenes.png" style="width:50.0%"></p>
</div>
</div>
<div id="part-11-hmrf-spatial-domains" class="section level2">
<h2 class="hasAnchor">
<a href="#part-11-hmrf-spatial-domains" class="anchor"></a>part 11: HMRF spatial domains</h2>
<div class="sourceCode" id="cb31"><pre class="downlit">
<span class="kw">hmrf_folder</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="kw">my_working_dir</span>,<span class="st">'/'</span>,<span class="st">'11_HMRF/'</span>)
<span class="co">if</span>(<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span>(<span class="kw">hmrf_folder</span>)) <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span>(<span class="kw">hmrf_folder</span>, recursive = <span class="kw">T</span>)

<span class="kw">my_spatial_genes</span> <span class="op">=</span> <span class="kw">km_spatialgenes</span>[<span class="fl">1</span><span class="op">:</span><span class="fl">100</span>]<span class="op">$</span><span class="kw">genes</span>

<span class="co"># do HMRF with different betas</span>
<span class="kw">HMRF_spatial_genes</span> <span class="op">=</span> <span class="fu"><a href="../reference/doHMRF.html">doHMRF</a></span>(gobject = <span class="kw">SS_seqfish</span>, 
                            expression_values = <span class="st">'scaled'</span>,
                            spatial_genes = <span class="kw">my_spatial_genes</span>,
                            spatial_network_name = <span class="st">'Delaunay_network'</span>,
                            k = <span class="fl">9</span>,
                            betas = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">28</span>,<span class="fl">2</span>,<span class="fl">3</span>), 
                            output_folder = <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="kw">hmrf_folder</span>, <span class="st">'/'</span>, <span class="st">'Spatial_genes/SG_top100_k9_scaled'</span>))


<span class="co">## view results of HMRF</span>
<span class="co">for</span>(<span class="kw">i</span> <span class="co">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="fl">28</span>, <span class="fl">32</span>, by = <span class="fl">2</span>)) {
  <span class="fu"><a href="../reference/viewHMRFresults2D.html">viewHMRFresults2D</a></span>(gobject = <span class="kw">SS_seqfish</span>,
                    HMRFoutput = <span class="kw">HMRF_spatial_genes</span>,
                    k = <span class="fl">9</span>, betas_to_view = <span class="kw">i</span>,
                    point_size = <span class="fl">2</span>)
}


<span class="co">## add HMRF of interest to giotto object</span>
<span class="kw">SS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="../reference/addHMRF.html">addHMRF</a></span>(gobject = <span class="kw">SS_seqfish</span>,
                  HMRFoutput = <span class="kw">HMRF_spatial_genes</span>,
                  k = <span class="fl">9</span>, betas_to_add = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">28</span>),
                  hmrf_name = <span class="st">'HMRF_2'</span>)


<span class="co">## visualize</span>
<span class="fu"><a href="../reference/spatPlot.html">spatPlot</a></span>(gobject = <span class="kw">SS_seqfish</span>, cell_color = <span class="st">'HMRF_2_k9_b.28'</span>, point_size = <span class="fl">3</span>, coord_fix_ratio = <span class="fl">1</span>, 
         save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'11_HMRF_2_k9_b.28'</span>, base_height = <span class="fl">3</span>, base_width = <span class="fl">9</span>, save_format = <span class="st">'pdf'</span>))
</pre></div>
</div>
<div id="part-12-cell-neighborhood-cell-typecell-type-interactions" class="section level2">
<h2 class="hasAnchor">
<a href="#part-12-cell-neighborhood-cell-typecell-type-interactions" class="anchor"></a>part 12: cell neighborhood: cell-type/cell-type interactions</h2>
<div class="sourceCode" id="cb32"><pre class="downlit">
<span class="kw">cell_proximities</span> <span class="op">=</span> <span class="fu"><a href="../reference/cellProximityEnrichment.html">cellProximityEnrichment</a></span>(gobject = <span class="kw">SS_seqfish</span>,
                                           cluster_column = <span class="st">'cell_types'</span>,
                                           spatial_network_name = <span class="st">'Delaunay_network'</span>,
                                           adjust_method = <span class="st">'fdr'</span>,
                                           number_of_simulations = <span class="fl">2000</span>)

<span class="co">## barplot</span>
<span class="fu"><a href="../reference/cellProximityBarplot.html">cellProximityBarplot</a></span>(gobject = <span class="kw">SS_seqfish</span>,
                     CPscore = <span class="kw">cell_proximities</span>, min_orig_ints = <span class="fl">5</span>, min_sim_ints = <span class="fl">5</span>, 
                     save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'12_a_barplot_cell_cell_enrichment'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/12_a_barplot_cell_cell_enrichment.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb33"><pre class="downlit">
<span class="co">## heatmap</span>
<span class="fu"><a href="../reference/cellProximityHeatmap.html">cellProximityHeatmap</a></span>(gobject = <span class="kw">SS_seqfish</span>, CPscore = <span class="kw">cell_proximities</span>, order_cell_types = <span class="kw">T</span>, scale = <span class="kw">T</span>,
                     color_breaks = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="op">-</span><span class="fl">1.5</span>, <span class="fl">0</span>, <span class="fl">1.5</span>), color_names = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'blue'</span>, <span class="st">'white'</span>, <span class="st">'red'</span>),
                     save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'12_b_heatmap_cell_cell_enrichment'</span>, unit = <span class="st">'in'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/12_b_heatmap_cell_cell_enrichment.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb34"><pre class="downlit">
<span class="co">## network</span>
<span class="fu"><a href="../reference/cellProximityNetwork.html">cellProximityNetwork</a></span>(gobject = <span class="kw">SS_seqfish</span>, CPscore = <span class="kw">cell_proximities</span>, remove_self_edges = <span class="kw">T</span>,
                     only_show_enrichment_edges = <span class="kw">T</span>,
                     save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'12_c_network_cell_cell_enrichment'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/12_c_network_cell_cell_enrichment.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb35"><pre class="downlit">
<span class="co">## network with self-edges</span>
<span class="fu"><a href="../reference/cellProximityNetwork.html">cellProximityNetwork</a></span>(gobject = <span class="kw">SS_seqfish</span>, CPscore = <span class="kw">cell_proximities</span>,
                     remove_self_edges = <span class="kw">F</span>, self_loop_strength = <span class="fl">0.3</span>,
                     only_show_enrichment_edges = <span class="kw">F</span>,
                     rescale_edge_weights = <span class="kw">T</span>,
                     node_size = <span class="fl">8</span>,
                     edge_weight_range_depletion = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">2</span>),
                     edge_weight_range_enrichment = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>,<span class="fl">5</span>),
                     save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'12_d_network_cell_cell_enrichment_self'</span>,
                                    base_height = <span class="fl">5</span>, base_width = <span class="fl">5</span>, save_format = <span class="st">'pdf'</span>))
</pre></div>
<p><embed src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/12_d_network_cell_cell_enrichment_self.pdf" style="width:50.0%"></embed></p>
<div class="sourceCode" id="cb36"><pre class="downlit">
<span class="co">## visualization of specific cell types</span>
<span class="co"># Option 1</span>
<span class="kw">spec_interaction</span> <span class="op">=</span> <span class="st">"astrocytes--Olig"</span>
<span class="fu"><a href="../reference/cellProximitySpatPlot2D.html">cellProximitySpatPlot2D</a></span>(gobject = <span class="kw">SS_seqfish</span>,
                        interaction_name = <span class="kw">spec_interaction</span>,
                        show_network = <span class="kw">T</span>,
                        cluster_column = <span class="st">'cell_types'</span>,
                        cell_color = <span class="st">'cell_types'</span>,
                        cell_color_code = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(astrocytes = <span class="st">'lightblue'</span>, Olig = <span class="st">'red'</span>),
                        point_size_select = <span class="fl">4</span>, point_size_other = <span class="fl">2</span>,
                        save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'12_e_cell_cell_enrichment_selected'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/12_e_cell_cell_enrichment_selected.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb37"><pre class="downlit">
<span class="co"># Option 2: create additional metadata</span>
<span class="kw">SS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="../reference/addCellIntMetadata.html">addCellIntMetadata</a></span>(<span class="kw">SS_seqfish</span>, 
                             spatial_network = <span class="st">'spatial_network'</span>,
                             cluster_column = <span class="st">'cell_types'</span>,
                             cell_interaction = <span class="kw">spec_interaction</span>,
                             name = <span class="st">'astro_olig_ints'</span>)
<span class="fu"><a href="../reference/spatPlot.html">spatPlot</a></span>(<span class="kw">SS_seqfish</span>, cell_color = <span class="st">'astro_olig_ints'</span>,
         select_cell_groups =  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'other_astrocytes'</span>, <span class="st">'other_Olig'</span>, <span class="st">'select_astrocytes'</span>, <span class="st">'select_Olig'</span>),
         legend_symbol_size = <span class="fl">3</span>, save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'12_f_cell_cell_enrichment_sel_vs_not'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/12_f_cell_cell_enrichment_sel_vs_not.png" style="width:50.0%"></p>
</div>
<div id="part-13-cell-neighborhood-interaction-changed-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#part-13-cell-neighborhood-interaction-changed-genes" class="anchor"></a>part 13: cell neighborhood: interaction changed genes</h2>
<div class="sourceCode" id="cb38"><pre class="downlit">
<span class="co">## select top 25th highest expressing genes</span>
<span class="kw">gene_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/fDataDT.html">fDataDT</a></span>(<span class="kw">SS_seqfish</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">gene_metadata</span><span class="op">$</span><span class="kw">nr_cells</span>, <span class="kw">gene_metadata</span><span class="op">$</span><span class="kw">mean_expr</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">gene_metadata</span><span class="op">$</span><span class="kw">nr_cells</span>, <span class="kw">gene_metadata</span><span class="op">$</span><span class="kw">mean_expr_det</span>)

<span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span>(<span class="kw">gene_metadata</span><span class="op">$</span><span class="kw">mean_expr_det</span>)
<span class="kw">high_expressed_genes</span> <span class="op">=</span> <span class="kw">gene_metadata</span>[<span class="kw">mean_expr_det</span> <span class="op">&gt;</span> <span class="fl">1.31</span>]<span class="op">$</span><span class="kw">gene_ID</span>

<span class="co">## identify genes that are associated with proximity to other cell types</span>
<span class="kw">CPGscoresHighGenes</span> <span class="op">=</span>  <span class="fu"><a href="../reference/findCPG.html">findCPG</a></span>(gobject = <span class="kw">SS_seqfish</span>,
                              selected_genes = <span class="kw">high_expressed_genes</span>,
                              spatial_network_name = <span class="st">'Delaunay_network'</span>,
                              cluster_column = <span class="st">'cell_types'</span>,
                              diff_test = <span class="st">'permutation'</span>,
                              adjust_method = <span class="st">'fdr'</span>,
                              nr_permutations = <span class="fl">2000</span>, 
                              do_parallel = <span class="kw">T</span>, cores = <span class="fl">4</span>)

<span class="co">## visualize all genes</span>
<span class="fu"><a href="../reference/plotCellProximityGenes.html">plotCellProximityGenes</a></span>(<span class="kw">SS_seqfish</span>, cpgObject = <span class="kw">CPGscoresHighGenes</span>, method = <span class="st">'dotplot'</span>, 
                       save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'13_a_CPG_dotplot'</span>, base_width = <span class="fl">5</span>, base_height = <span class="fl">5</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/13_a_CPG_dotplot.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb39"><pre class="downlit">
<span class="co">## filter genes</span>
<span class="kw">CPGscoresFilt</span> <span class="op">=</span> <span class="fu"><a href="../reference/filterCPG.html">filterCPG</a></span>(<span class="kw">CPGscoresHighGenes</span>)

<span class="co">## visualize subset of interaction changed genes (ICGs)</span>
<span class="kw">ICG_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Jakmip1'</span>, <span class="st">'Golgb1'</span>, <span class="st">'Dact2'</span>, <span class="st">'Ddx27'</span>, <span class="st">'Abl1'</span>, <span class="st">'Zswim8'</span>)
<span class="kw">ICG_genes_types</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Lhx6 iNeuron'</span>, <span class="st">'Lhx6 iNeuron'</span>, <span class="st">'L4 eNeuron'</span>, <span class="st">'L4 eNeuron'</span>, <span class="st">'astrocytes'</span>, <span class="st">'astrocytes'</span>)
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw">ICG_genes</span>) <span class="op">=</span> <span class="kw">ICG_genes_types</span>

<span class="fu"><a href="../reference/plotICG.html">plotICG</a></span>(gobject = <span class="kw">SS_seqfish</span>,
        cpgObject = <span class="kw">CPGscoresHighGenes</span>,
        source_type = <span class="st">'endothelial'</span>,
        source_markers = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Pltp'</span>, <span class="st">'Cldn5'</span>, <span class="st">'Apcdd1'</span>),
        ICG_genes = <span class="kw">ICG_genes</span>,
        save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'13_b_ICG_barplot'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/13_b_ICG_barplot.png" style="width:50.0%"></p>
</div>
<div id="part-14-cell-neighborhood-ligand-receptor-cell-cell-communication" class="section level2">
<h2 class="hasAnchor">
<a href="#part-14-cell-neighborhood-ligand-receptor-cell-cell-communication" class="anchor"></a>part 14: cell neighborhood: ligand-receptor cell-cell communication</h2>
<div class="sourceCode" id="cb40"><pre class="downlit">
<span class="co"># LR expression</span>
<span class="co"># LR activity changes</span>
<span class="kw">LR_data</span> <span class="op">=</span> <span class="kw">data.table</span>::<span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"mouse_ligand_receptors.txt"</span>, package = <span class="st">'Giotto'</span>))

<span class="kw">LR_data</span>[, <span class="kw">ligand_det</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="kw">mouseLigand</span> <span class="op">%in%</span> <span class="kw">SS_seqfish</span><span class="op">@</span>gene_ID, <span class="kw">T</span>, <span class="kw">F</span>)]
<span class="kw">LR_data</span>[, <span class="kw">receptor_det</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="kw">mouseReceptor</span> <span class="op">%in%</span> <span class="kw">SS_seqfish</span><span class="op">@</span>gene_ID, <span class="kw">T</span>, <span class="kw">F</span>)]
<span class="kw">LR_data_det</span> <span class="op">=</span> <span class="kw">LR_data</span>[<span class="kw">ligand_det</span> <span class="op">==</span> <span class="kw">T</span> <span class="op">&amp;</span> <span class="kw">receptor_det</span> <span class="op">==</span> <span class="kw">T</span>]
<span class="kw">select_ligands</span> <span class="op">=</span> <span class="kw">LR_data_det</span><span class="op">$</span><span class="kw">mouseLigand</span>
<span class="kw">select_receptors</span> <span class="op">=</span> <span class="kw">LR_data_det</span><span class="op">$</span><span class="kw">mouseReceptor</span>


<span class="co">## get statistical significance of gene pair expression changes based on expression ##</span>
<span class="kw">expr_only_scores</span> <span class="op">=</span> <span class="fu"><a href="../reference/exprCellCellcom.html">exprCellCellcom</a></span>(gobject = <span class="kw">SS_seqfish</span>,
                                   cluster_column = <span class="st">'cell_types'</span>, 
                                   random_iter = <span class="fl">1000</span>,
                                   gene_set_1 = <span class="kw">select_ligands</span>,
                                   gene_set_2 = <span class="kw">select_receptors</span>, 
                                   verbose = <span class="fl">FALSE</span>)

<span class="co">## get statistical significance of gene pair expression changes upon cell-cell interaction</span>
<span class="kw">spatial_all_scores</span> <span class="op">=</span> <span class="fu"><a href="../reference/spatCellCellcom.html">spatCellCellcom</a></span>(<span class="kw">SS_seqfish</span>,
                                     spatial_network_name = <span class="st">'spatial_network'</span>,
                                     cluster_column = <span class="st">'cell_types'</span>, 
                                     random_iter = <span class="fl">1000</span>,
                                     gene_set_1 = <span class="kw">select_ligands</span>,
                                     gene_set_2 = <span class="kw">select_receptors</span>,
                                     adjust_method = <span class="st">'fdr'</span>,
                                     do_parallel = <span class="kw">T</span>,
                                     cores = <span class="fl">4</span>,
                                     verbose = <span class="st">'a little'</span>)


<span class="co">## select top LR ##</span>
<span class="kw">selected_spat</span> <span class="op">=</span> <span class="kw">spatial_all_scores</span>[<span class="kw">p.adj</span> <span class="op">&lt;=</span> <span class="fl">0.01</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="kw">log2fc</span>) <span class="op">&gt;</span> <span class="fl">0.25</span> <span class="op">&amp;</span> <span class="kw">lig_nr</span> <span class="op">&gt;=</span> <span class="fl">4</span> <span class="op">&amp;</span> <span class="kw">rec_nr</span> <span class="op">&gt;=</span> <span class="fl">4</span>]
<span class="kw">data.table</span>::<span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setorder.html">setorder</a></span>(<span class="kw">selected_spat</span>, <span class="op">-</span><span class="kw">PI</span>)

<span class="kw">top_LR_ints</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="kw">selected_spat</span>[<span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="kw">PI</span>))]<span class="op">$</span><span class="kw">LR_comb</span>)[<span class="fl">1</span><span class="op">:</span><span class="fl">33</span>]
<span class="kw">top_LR_cell_ints</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="kw">selected_spat</span>[<span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="kw">PI</span>))]<span class="op">$</span><span class="kw">LR_cell_comb</span>)[<span class="fl">1</span><span class="op">:</span><span class="fl">33</span>]

<span class="fu"><a href="../reference/plotCCcomDotplot.html">plotCCcomDotplot</a></span>(gobject = <span class="kw">SS_seqfish</span>,
                 comScores = <span class="kw">spatial_all_scores</span>,
                 selected_LR = <span class="kw">top_LR_ints</span>,
                 selected_cell_LR = <span class="kw">top_LR_cell_ints</span>,
                 cluster_on = <span class="st">'PI'</span>,
                 save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'14_a_communication_dotplot'</span>, save_format = <span class="st">'pdf'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/14_a_communication_dotplot.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb41"><pre class="downlit">
<span class="co">## spatial vs rank ####</span>
<span class="kw">comb_comm</span> <span class="op">=</span> <span class="fu"><a href="../reference/combCCcom.html">combCCcom</a></span>(spatialCC = <span class="kw">spatial_all_scores</span>,
                      exprCC = <span class="kw">expr_only_scores</span>)


<span class="co"># highest levels of ligand and receptor prediction</span>
<span class="co"># top differential activity levels for ligand receptor pairs</span>
<span class="fu"><a href="../reference/plotRankSpatvsExpr.html">plotRankSpatvsExpr</a></span>(gobject = <span class="kw">SS_seqfish</span>,
                   <span class="kw">comb_comm</span>,
                   expr_rnk_column = <span class="st">'LR_expr_rnk'</span>,
                   spat_rnk_column = <span class="st">'LR_spat_rnk'</span>,
                   midpoint = <span class="fl">10</span>,
                   save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'14_b_expr_vs_spatial_expression_rank'</span>,
                                  base_height = <span class="fl">4</span>, base_width = <span class="fl">4.5</span>, save_format = <span class="st">'pdf'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/14_b_expr_vs_spatial_expression_rank.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb42"><pre class="downlit">
<span class="co"># recovery</span>
<span class="fu"><a href="../reference/plotRecovery.html">plotRecovery</a></span>(gobject = <span class="kw">SS_seqfish</span>,
             <span class="kw">comb_comm</span>,
             expr_rnk_column = <span class="st">'LR_expr_rnk'</span>,
             spat_rnk_column = <span class="st">'LR_spat_rnk'</span>,
             ground_truth = <span class="st">'spatial'</span>,
             save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'14_c_spatial_recovery_expression_rank'</span>, 
                            base_height = <span class="fl">3</span>, base_width = <span class="fl">3</span>, save_format = <span class="st">'pdf'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/14_c_spatial_recovery_expression_rank.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb43"><pre class="downlit">
<span class="co"># highest differential activity of ligand and receptor prediction</span>

<span class="co"># top differential activity levels for ligand receptor pairs</span>
<span class="fu"><a href="../reference/plotRankSpatvsExpr.html">plotRankSpatvsExpr</a></span>(gobject = <span class="kw">SS_seqfish</span>,
                   <span class="kw">comb_comm</span>,
                   expr_rnk_column = <span class="st">'exprPI_rnk'</span>,
                   spat_rnk_column = <span class="st">'spatPI_rnk'</span>,
                   midpoint = <span class="fl">10</span>,
                   save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'14_d_expr_vs_spatial_activity'</span>,
                                  base_height = <span class="fl">4</span>, base_width = <span class="fl">4.5</span>, save_format = <span class="st">'pdf'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/14_d_expr_vs_spatial_activity.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb44"><pre class="downlit">
<span class="fu"><a href="../reference/plotRecovery.html">plotRecovery</a></span>(gobject = <span class="kw">SS_seqfish</span>,
             <span class="kw">comb_comm</span>,
             expr_rnk_column = <span class="st">'exprPI_rnk'</span>,
             spat_rnk_column = <span class="st">'spatPI_rnk'</span>,
             ground_truth = <span class="st">'spatial'</span>,
             save_param = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(save_name = <span class="st">'14_e_spatial_recovery_activity'</span>, 
                            base_height = <span class="fl">3</span>, base_width = <span class="fl">3</span>, save_format = <span class="st">'pdf'</span>))
</pre></div>
<p><img src="../inst/images/mouse_seqFISH_SS_cortex/vignette_200914/14_e_spatial_recovery_activity.png" style="width:50.0%"></p>
</div>
<div id="part-15-export-giotto-analyzer-to-viewer" class="section level2">
<h2 class="hasAnchor">
<a href="#part-15-export-giotto-analyzer-to-viewer" class="anchor"></a>part 15: export Giotto Analyzer to Viewer</h2>
<div class="sourceCode" id="cb45"><pre class="downlit">
<span class="kw">viewer_folder</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="kw">my_working_dir</span>, <span class="st">'/'</span>, <span class="st">'Mouse_cortex_viewer'</span>)

<span class="co"># select annotations, reductions and expression values to view in Giotto Viewer</span>
<span class="fu"><a href="../reference/pDataDT.html">pDataDT</a></span>(<span class="kw">SS_seqfish</span>)
<span class="fu"><a href="../reference/exportGiottoViewer.html">exportGiottoViewer</a></span>(gobject = <span class="kw">SS_seqfish</span>, output_directory = <span class="kw">viewer_folder</span>,
                   factor_annotations = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'cell_types'</span>,
                                          <span class="st">'leiden_clus'</span>,
                                          <span class="st">'sub_leiden_clus_select'</span>,
                                          <span class="st">'HMRF_2_k9_b.28'</span>),
                   numeric_annotations = <span class="st">'total_expr'</span>,
                   dim_reductions = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'umap'</span>),
                   dim_reduction_names = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'umap'</span>),
                   expression_values = <span class="st">'scaled'</span>,
                   expression_rounding = <span class="fl">3</span>,
                   overwrite_dir = <span class="fl">TRUE</span>)
</pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://www.drieslab.com/">Ruben Dries</a>, Qian Zhu, Huipeng Li, Rui Dong, Guo-Cheng Yuan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
