<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>findGiniMarkers — findGiniMarkers • Giotto</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="findGiniMarkers — findGiniMarkers" />
<meta property="og:description" content="Identify marker genes for selected clusters based on gini detection and expression scores." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-109153024-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-109153024-2');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Giotto</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Start
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/getting_started.html#description">Description</a>
    </li>
    <li>
      <a href="../articles/getting_started.html#workflow-diagram">Workflow</a>
    </li>
    <li>
      <a href="../articles/getting_started.html#howtos">HOWTOs</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Documentation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Datasets
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mouse_seqFISH_cortex_200319.html">mouse seqFISH cortex</a>
    </li>
    <li>
      <a href="../articles/mouse_merFISH_hypoth_preoptic_region_200326.html">mouse hypoth. preopt. region</a>
    </li>
    <li>
      <a href="../articles/mouse_starmap_cortex.html">mouse STARmap cortex</a>
    </li>
    <li>
      <a href="../articles/human_cycif_PDAC_200322.html">human CyCIF PDAC</a>
    </li>
    <li>
      <a href="../articles/mouse_visium_brain_200325.html">mouse Visium brain</a>
    </li>
    <li>
      <a href="../articles/mouse_visium_kidney_200326.html">mouse Visium kidney</a>
    </li>
    <li>
      <a href="../articles/mouse_CODEX_spleen.html">mouse CODEX spleen</a>
    </li>
    <li>
      <a href="../articles/mouse_osmFISH_SScortex.html">mouse osmFISH SScortex</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/faqs.html">FAQ</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/RubD/Giotto">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>findGiniMarkers</h1>
    <small class="dont-index">Source: <a href='https://github.com/RubD/Giotto/blob/master/R/differential_expression.R'><code>R/differential_expression.R</code></a></small>
    <div class="hidden name"><code>findGiniMarkers.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Identify marker genes for selected clusters based on gini detection and expression scores.</p>
    </div>

    <pre class="usage"><span class='fu'>findGiniMarkers</span>(
  <span class='no'>gobject</span>,
  <span class='kw'>expression_values</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"normalized"</span>, <span class='st'>"scaled"</span>, <span class='st'>"custom"</span>),
  <span class='no'>cluster_column</span>,
  <span class='kw'>subset_clusters</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>group_1</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>group_2</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>min_expr_gini_score</span> <span class='kw'>=</span> <span class='fl'>0.2</span>,
  <span class='kw'>min_det_gini_score</span> <span class='kw'>=</span> <span class='fl'>0.2</span>,
  <span class='kw'>detection_threshold</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>rank_score</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>min_genes</span> <span class='kw'>=</span> <span class='fl'>5</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gobject</th>
      <td><p>giotto object</p></td>
    </tr>
    <tr>
      <th>expression_values</th>
      <td><p>gene expression values to use</p></td>
    </tr>
    <tr>
      <th>cluster_column</th>
      <td><p>clusters to use</p></td>
    </tr>
    <tr>
      <th>subset_clusters</th>
      <td><p>selection of clusters to compare</p></td>
    </tr>
    <tr>
      <th>group_1</th>
      <td><p>group 1 cluster IDs from cluster_column for pairwise comparison</p></td>
    </tr>
    <tr>
      <th>group_2</th>
      <td><p>group 2 cluster IDs from cluster_column for pairwise comparison</p></td>
    </tr>
    <tr>
      <th>min_expr_gini_score</th>
      <td><p>filter on minimum gini coefficient for expression</p></td>
    </tr>
    <tr>
      <th>min_det_gini_score</th>
      <td><p>filter on minimum gini coefficient for detection</p></td>
    </tr>
    <tr>
      <th>detection_threshold</th>
      <td><p>detection threshold for gene expression</p></td>
    </tr>
    <tr>
      <th>rank_score</th>
      <td><p>rank scores for both detection and expression to include</p></td>
    </tr>
    <tr>
      <th>min_genes</th>
      <td><p>minimum number of top genes to return</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>data.table with marker genes</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Detection of marker genes using the <a href='https://en.wikipedia.org/wiki/Gini_coefficient'>https://en.wikipedia.org/wiki/Gini_coefficient</a>gini
coefficient is based on the following steps/principles per gene:</p><ul>
<li><p>1. calculate average expression per cluster</p></li>
<li><p>2. calculate detection fraction per cluster</p></li>
<li><p>3. calculate gini-coefficient for av. expression values over all clusters</p></li>
<li><p>4. calculate gini-coefficient for detection fractions over all clusters</p></li>
<li><p>5. convert gini-scores to rank scores</p></li>
<li><p>6. for each gene create combined score = detection rank x expression rank x expr gini-coefficient x detection gini-coefficient</p></li>
<li><p>7. for each gene sort on expression and detection rank and combined score</p></li>
</ul>

<p>As a results "top gini" genes are genes that are very selectivily expressed in a specific cluster,
however not always expressed in all cells of that cluster. In other words highly specific, but
not necessarily sensitive at the single-cell level.</p>
<p>To perform differential expression between cluster groups you need to specificy cluster IDs
to the parameters <em>group_1</em> and <em>group_2</em>.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples">    <span class='fu'>findGiniMarkers</span>(<span class='no'>gobject</span>)</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://www.rubendries.com/'>Ruben Dries</a>, Huipeng Li, Qian Zhu, Rui Dong, Guo-Cheng Yuan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


